raster/rt_pg/rtpg_geometry.c

```diff
@@ -651,6 +654,324 @@ Datum RASTER_getPixelPolygons(PG_FUNCTION_ARGS)
+			rt_raster_destroy(raster);
+			PG_FREE_IF_COPY(pgraster, 0);
+			MemoryContextSwitchTo(oldcontext);
+			SRF_RETURN_DONE(funcctx);
+		}
+
+		/* band specified, load band and info */
+		if (hasband) {
+			numbands = rt_raster_get_num_bands(raster);
+			POSTGIS_RT_DEBUGF(3, "band %d", nband);
+			POSTGIS_RT_DEBUGF(3, "# of bands %d", numbands);
+
+			/* check band index */
+			if (nband < 1 || nband > numbands) {
+				elog(NOTICE, "Invalid band index (must use 1-based). Returning NULL");
+				rt_raster_destroy(raster);
+				PG_FREE_IF_COPY(pgraster, 0);
+				MemoryContextSwitchTo(oldcontext);
+				SRF_RETURN_DONE(funcctx);
+			}
+
+			/* get band */
+			band = rt_raster_get_band(raster, nband - 1);
+			if (!band) {
+				elog(NOTICE, "Could not find band at index %d. Returning NULL", nband);
+
+				rt_raster_destroy(raster);
+				PG_FREE_IF_COPY(pgraster, 0);
+
+				MemoryContextSwitchTo(oldcontext);
+				SRF_RETURN_DONE(funcctx);
+			}
+
+			if (!rt_band_get_hasnodata_flag(band))
+				exclude_nodata_value = FALSE;
+		}
+
+		/* set bounds if columnx, rowy not set */
+		if (nocolumnx) {
+			bounds[0] = 1;
+			bounds[1] = rt_raster_get_width(raster);
+		}
+		if (norowy) {
+			bounds[2] = 1;
+			bounds[3] = rt_raster_get_height(raster);
+		}
+		POSTGIS_RT_DEBUGF(3, "bounds (min x, max x, min y, max y) = (%d, %d, %d, %d)",
+			bounds[0], bounds[1], bounds[2], bounds[3]);
+
+		/* rowy */
+		pixcount = 0;
+		for (y = bounds[2]; y <= bounds[3]; y++) {
+			/* columnx */
+			for (x = bounds[0]; x <= bounds[1]; x++) {
+
+				value = 0;
+				isnodata = TRUE;
+
+				if (hasband) {
+					if (rt_band_get_pixel(band, x - 1, y - 1, &value, &isnodata) != ES_NONE) {
+						/* free already created centroid points */
+						for (i = 0; i < pixcount; i++)
+							lwgeom_free(pix[i].geom);
+						if (pixcount) pfree(pix);
+
+						rt_band_destroy(band);
+						rt_raster_destroy(raster);
+						PG_FREE_IF_COPY(pgraster, 0);
+
+						MemoryContextSwitchTo(oldcontext);
+						elog(ERROR, "RASTER_getPixelCentroids: Could not get pixel value");
+						SRF_RETURN_DONE(funcctx);
+					}
+
+					/* don't continue if pixel is NODATA and to exclude NODATA */
+					if (isnodata && exclude_nodata_value) {
+						POSTGIS_RT_DEBUG(5, "pixel value is NODATA and exclude_nodata_value = TRUE");
+						continue;
+					}
+				}
+
+				/* geometry */
+				point = rt_raster_pixel_as_centroid_point(raster, x - 1, y - 1);
+				if (!point) {
+					/*  free already created centroid points */
+					for (i = 0; i < pixcount; i++)
+						lwgeom_free(pix[i].geom);
+					if (pixcount) pfree(pix);
+
+					if (hasband) rt_band_destroy(band);
+					rt_raster_destroy(raster);
+					PG_FREE_IF_COPY(pgraster, 0);
+
+					MemoryContextSwitchTo(oldcontext);
+					elog(ERROR, "RASTER_getPixelCentroids: Could not get pixel centroid");
+					SRF_RETURN_DONE(funcctx);
+				}
+
+				/* allocate space for new point */
+				if (!pixcount) {
+					pix = palloc(sizeof(struct rt_pixel_t) * (pixcount + 1));
+				} {
```

Comment:
this anonymous block looks sus. repalloc right after palloc?

Reply:
Fixed. Good thing you caught it.